### PR TITLE
status identifies failing component, fleet gateway may report degraded, liveness endpoint added

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Extend the status reporter to include status update time, and the message
from the failing component or app.

Allow the fleet-gateway (component) to report a degraded state if two checkins
fail.

Add /liveness route to metrics server. This route will report the status. A 200 is
returned if the agent is healthy, and a 503 is returned if it is degraded or failed,
the response body is a json object with additional details. Status may not
propagate to fleet-server in the case where degradation is caused by checkin
fails. It may also lead to the server reporting degraded for up to 30s (fleet-server
polling time) when the agent is able to successfully connect.

## Why is it important?

Allowing the agent to report a degraded state when it cannot contact fleet is
useful as it can be used to better debug connectivity issues. Exposing this
information on a liveness endpoint enables platforms (such as k8s) to poll it
and manage the lifecycle of the agent.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- Build the agent locally
- Run the elastic-stack with `elastic-package stack up`
- ensure that the agent + fleet server appears in the fleet ui
- stop the fleet-server container
- wait a minute then ensure that elastic-agent container reports a degraded status locally
- `curl` the liveness endpoint and ensure a 503 is returned

## Related issues

- Relates #390

### CODE REVIEW

File: internal/pkg/agent/application/gateway/fleet/noop_status_controller.go [hunk 1]
```diff
@@ -5,21 +5,25 @@
...
-func (*noopController) Status() status.AgentStatus                     { return status.AgentStatus{Status: status.Healthy} }
-func (*noopController) StatusCode() status.AgentStatusCode             { return status.Healthy }
-func (*noopController) UpdateStateID(_ string)                         {}
-func (*noopController) StatusString() string                           { return "online" }
+func (*noopController) RegisterApp(_ string, _ string) status.Reporter   { return &noopReporter{} }
+func (*noopController) Status() status.AgentStatus                       { return status.AgentStatus{Status: status.Healthy} }
+func (*noopController) StatusCode() status.AgentStatusCode               { return status.Healthy }
+func (*noopController) UpdateStateID(_ string)                           {}
+func (*noopController) StatusString() string                             { return "online" }
+func (*noopController) ServeHTTP(_ http.ResponseWriter, _ *http.Request) {}
```
[the remark this answers is not in the document]
Comment (ph) [reply internal/pkg/agent/application/gateway/fleet/noop_status_controller.go:27]: Yeah that would make sense +1 to rename.

File: internal/pkg/core/status/handler.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		status = http.StatusServiceUnavailable
+	}
+
+	wr.Header().Set("Content-Type", "application/json")
+	wr.WriteHeader(status)
+	enc := json.NewEncoder(wr)
+	if err := enc.Encode(lr); err != nil {
+		r.log.Errorf("Unable to encode liveness response: %v", err)
+	}
+}
```
Comment (ph) [inline internal/pkg/core/status/handler.go:39]: @michel-laterman When we add the new information for the liveness check we refactor the handler do a dependency injection style and have a `StatusReporter` that take a an interface to query for status?

**Review by ph — APPROVED** [review]

**Comment by michel-laterman** [comment]
👍 for adding an e2e test

File: internal/pkg/core/status/handler.go (hunk 2 above)
Comment (michel-laterman) [reply internal/pkg/core/status/handler.go:39]: Yeah, this is probably going to be the case for the v2 implementation.

**Pull request merged**